model/model.js

```diff
@@ -21,30 +21,67 @@ exports.retrieveCommentsByArticleId = (articleId) => {
     });
 };
 
-exports.retrieveAllArticles = () => {
-  return db
-    .query(
-      `SELECT
-    a.article_id,
-    a.title,
-    a.topic,
-    a.author AS author,
-    a.created_at,
-    a.votes AS votes,
-    a.article_img_url,
-    COALESCE(COUNT(c.comment_id), 0) AS comment_count
-FROM
-    articles a
-LEFT JOIN
-    comments c ON a.article_id = c.article_id
-GROUP BY
-    a.article_id
-ORDER BY
-    a.created_at DESC;`
-    )
-    .then(({ rows }) => {
+exports.retrieveAllArticles = (topic) => {
+  let query = `
+    SELECT
+      a.article_id,
+      a.title,
+      a.topic,
+      a.author AS author,
+      a.created_at,
+      a.votes AS votes,
+      a.article_img_url,
+      COALESCE(COUNT(c.comment_id), 0) AS comment_count
+    FROM
+      articles a
+    LEFT JOIN
+      comments c ON a.article_id = c.article_id
+    `;
+
+  const values = [];
+
+  if (topic) {
+    query += `WHERE a.topic = $1`;
+    values.push(topic);
+  }
+
+  query += `
+    GROUP BY
+      a.article_id
+    ORDER BY
+      a.created_at DESC;
+  `;
+
+  if (!topic) {
+    return db.query(query, values).then(({ rows }) => {
       return rows;
     });
+  }
+
+  const getAllTopicsQuery = `SELECT slug FROM topics`;
+
+  return Promise.all([
+    db.query(query, values),
+    db.query(getAllTopicsQuery),
+  ]).then(([articles, topics]) => {
+    const validTopic = topics.rows.some(
+      (topicItem) => topicItem.slug === topic
+    );
+
+    if (!validTopic) {
+      return Promise.reject({ status: 404, msg: "404 not found" });
+    }
+
+    if (articles.rows.length === 0 && validTopic) {
+      return Promise.resolve({ status: 200, msg: [] });
+    }
+
+    if (articles.rows.length === 0) {
+      return Promise.reject({ status: 404, msg: "404 not found" });
+    }
```

Comment:
A few notes about this section:
You're checking if the topic is valid on line 71, so there's no way that validTopic could be false past that line, so there's no point in checking it on line 75.

Also, instead of sending back a promise.resolve, it would make more sense to let the rows be sent back normally by getting to your return on line 83 - since that's not an error case.

Remember that the only error for this endpoint is a non existing topic! An empty array for articles is OK - so it seems like you can get rid of the last 2 if statements in this section.